packages/react-shopify-app-bridge/src/Provider.tsx

```diff
@@ -1,160 +1,150 @@
+import { getSessionToken } from "@shopify/app-bridge-utils";
 import { Redirect } from "@shopify/app-bridge/actions";
-import assert from "assert";
-import { isArray, isString, isUndefined } from "lodash";
-import React, { memo, useEffect, useMemo, useState } from "react";
-import { useRender } from "./hooks";
-import {
-  GadgetAuthContext,
-  GadgetContext,
-  generateCodeChallenge,
-  generateCodeVerifier,
-  getGadgetAccessToken,
-  useLocalStorage,
-  useSessionStorage,
-} from "./index";
-import { getTokenKey } from "./utils";
+import { isUndefined } from "lodash";
+import React, { memo, useContext, useEffect, useMemo, useState } from "react";
+import { useQuery } from "urql";
+import { GadgetAuthContext, GadgetAuthContextValue } from "./index";
 
 export enum AppType {
   Standalone,
   Embedded,
 }
 
+/** Internal props used to create the right structure of providers */
 type GadgetProviderProps = {
   children: JSX.Element | JSX.Element[];
-  appBridge?: ClientApplication<AppBridgeState>;
   forceRedirect: boolean;
-  runningShopifyAuth: boolean;
-  runningGadgetAuth: boolean;
-  setAuth: (value: string) => void;
   isEmbedded: boolean;
   gadgetAppUrl: string;
-  token?: string;
   originalQueryParams?: URLSearchParams;
+  api: AnyClient;
 };
 
-const gadgetAuthParams = ["authorizationCode"];
+const GetCurrentSessionQuery = `
+  query GetSessionForShopifyApp {
+    currentSession {
+      id
+      shop {
+        id
+      }
+    }
+  }
+`;
 
-const InnerGadgetProvider = memo(
-  ({
-    children,
+// inner component that exists in order to ask for the app bridge
+const InnerGadgetProvider = memo(({ children, forceRedirect, isEmbedded, gadgetAppUrl, originalQueryParams, api }: GadgetProviderProps) => {
+  const appBridge = useContext(AppBridgeContext);
+
+  const [context, setContext] = useState<GadgetAuthContextValue>({
+    isAuthenticated: false,
+    isEmbedded: false,
+    canAuth: false,
+    loading: false,
     appBridge,
-    forceRedirect,
-    runningShopifyAuth,
-    runningGadgetAuth,
-    setAuth,
-    isEmbedded,
-    gadgetAppUrl,
-    token,
-    originalQueryParams,
-  }: GadgetProviderProps) => {
-    const [context, setContext] = useState<GadgetContext>({
-      isAuthenticated: false,
-      isEmbedded: false,
-      canAuth: false,
-      loading: false,
-      appBridge,
-    });
+  });
 
-    useEffect(() => {
-      if (!runningShopifyAuth) return;
+  useEffect(() => {
+    if (!appBridge) return;
+    // setup the api client to always query using the custom shopify auth implementation
+    api.connection.setAuthenticationMode({
+      custom: {
+        processFetch: async (_input, init) => {
+          const headers = new Headers(init.headers);
+          headers.append("Authorization", `ShopifySessionToken ${await getSessionToken(appBridge)}`);
+          init.headers ??= {};
+          headers.forEach(function (value, key) {
+            (init.headers as Record<string, string>)[key] = value;
```

Comment:
Is `init.headers` being used externally, after this mutation? I'm just curious because we don't seem to be using `init` again (after line 56).

Reply:
Yeah, it is, this is a kinda weird function that mutates the stuff we're about to pass to a `fetch` function. See https://github.com/gadget-inc/js-clients/blob/main/packages/api-client-core/src/GadgetConnection.ts#L250